go/merkle/encoder.go

```diff
@@ -0,0 +1,71 @@
+package merkle
+
+import (
+	"crypto/hmac"
+	cryptorand "crypto/rand"
+	"crypto/sha256"
+	"crypto/sha512"
+
+	"github.com/pkg/errors"
+)
+
+type Encoder struct {
+	encodingType EncodingType
+}
+
+func NewEncoder(encodingType EncodingType) *Encoder {
+	return &Encoder{encodingType: encodingType}
+}
+
+func (e *Encoder) BlindedPreimage(leaf Leaf, key Key, secret Secret) (BlindedPreimage, error) {
+	switch e.encodingType {
+	case EncodingTypeBlindedSHA256v1:
+		h := hmac.New(sha256.New, secret.Secret)
+		h.Write(key.Key)
+		return NewBlindedPreimage(leaf, h.Sum(nil))
+	case EncodingTypeBlindedSHA512_256v1:
+		h := hmac.New(sha512.New512_256, secret.Secret)
+		h.Write(key.Key)
+		return NewBlindedPreimage(leaf, h.Sum(nil))
+	default:
+		return BlindedPreimage{}, errors.Errorf("unknown encoding type %q", e.encodingType)
+	}
+}
+
+func (e *Encoder) Hash(preimage BlindedPreimage) ([]byte, error) {
+	b, err := preimage.LeafContainer.Serialize()
+	if err != nil {
+		return nil, err
+	}
+	switch e.encodingType {
+	case EncodingTypeBlindedSHA256v1:
+		h := hmac.New(sha256.New, preimage.BlindedEntropy)
+		h.Write(b)
+		return h.Sum(nil), nil
+	case EncodingTypeBlindedSHA512_256v1:
+		h := hmac.New(sha512.New512_256, preimage.BlindedEntropy)
```

Comment:
elsewhere we've been using HMAC-SHA512 truncated to the first 32-bytes, which is different. Can we use that here to for HMAC?  For hashing below we can use 512/256

Reply:
ok, so use sha512[:32] for the blinding hashes, but sha512/256 for the tree hashes?

Reply:
I think this would be more consistent with what we're doing now. Thanks